File: laboratorio3/planets.py
import heapq

n, m = map(int,input().split())
grafo = {i:{} for i in range(1, n + 1)}
custos = {i: float("inf") for i in range(1, n + 1)}
custos[1] = 0
restricoes = {i:set() for i in range(1 , n + 1)}

for _ in range(m):
    a, b, c = map(int,input().split())
    grafo[a][b] = c
    grafo[b][a] = c

for i in range(1, n + 1):
    partes = input().split()
    
    k = int(partes[0])
    if k > 0:
        tempos = map(int, partes[1:])
        restricoes[i] = set(tempos)

fila = [(0, 1)]

while fila:
    custo_atual, no_atual = heapq.heappop(fila)
    if custo_atual > custos[no_atual]:
        continue
    tempo_de_partida = custo_atual
    
    while tempo_de_partida in restricoes[no_atual]:
        tempo_de_partida += 1

    for vizinho, custo_viagem in grafo[no_atual].items():
        novo_custo_chegada = tempo_de_partida + custo_viagem
        
        if novo_custo_chegada < custos[vizinho]:
            custos[vizinho] = novo_custo_chegada
            heapq.heappush(fila, (novo_custo_chegada, vizinho))

print(custos[n]) if custos[n] != float("inf") else print(-1)

File: laboratorio3/planets.cpp
#include <iostream>
#include <vector>
#include <map>
#include <unordered_set>
#include <queue>
#include <climits>
#include <string>
#include <sstream>

using ll = long long;
using pii = std::pair<ll, int>;

int main() {
    std::ios_base::sync_with_stdio(false);
    std::cin.tie(NULL);

    int n, m;
    std::cin >> n >> m;
    std::vector<std::map<int, int>> grafo(n + 1);
    std::vector<ll> custos(n + 1, LLONG_MAX);
    custos[1] = 0;

    std::vector<std::unordered_set<int>> restricoes(n + 1);

    for (int i = 0; i < m; ++i) {
        int a, b, c;
        std::cin >> a >> b >> c;
        grafo[a][b] = c;
        grafo[b][a] = c;
    }

    std::cin.ignore(); 
    
    for (int i = 1; i <= n; ++i) {
        std::string linha;
        std::getline(std::cin, linha);
        std::stringstream ss(linha);

        int k;
        ss >> k;
        if (k > 0) {
            int tempo;
            while (ss >> tempo) {
                restricoes[i].insert(tempo);
            }
        }
    }

    std::priority_queue<pii, std::vector<pii>, std::greater<pii>> fila;
    fila.push({0, 1});

    while (!fila.empty()) {
        ll custo_atual = fila.top().first;
        int no_atual = fila.top().second;
        fila.pop();

        if (custo_atual > custos[no_atual]) {
            continue;
        }

        ll tempo_de_partida = custo_atual;
        
        while (restricoes[no_atual].count(tempo_de_partida)) {
            tempo_de_partida++;
        }

        for (auto const& [vizinho, custo_viagem] : grafo[no_atual]) {
            ll novo_custo_chegada = tempo_de_partida + custo_viagem;

            if (novo_custo_chegada < custos[vizinho]) {
                custos[vizinho] = novo_custo_chegada;
                fila.push({novo_custo_chegada, vizinho});
            }
        }
    }

    if (custos[n] == LLONG_MAX) {
        std::cout << -1 << std::endl;
    } else {
        std::cout << custos[n] << std::endl;
    }

    return 0;
}